src/hooks/useMultiNetworkAddressBalances.ts

```diff
@@ -0,0 +1,123 @@
+
+interface useMultiNetworkAddressBalancesArgs {
+  ownerAddress: string | undefined
+  tokenAddress: 'NATIVE' | string | undefined
+}
+
+type AddressNetworkBalanceData = Partial<
+  Record<
+    SupportedChainId,
+    Record<string | 'NATIVE', CurrencyAmount<Token> | CurrencyAmount<NativeCurrency> | undefined>
+  >
+>
+
+interface handleBalanceArg {
+  amount: CurrencyAmount<Currency>
+  chainId: SupportedChainId
+  tokenAddress: string | 'NATIVE'
+}
+
+const testnetSet = new Set(TESTNET_CHAIN_IDS) as Set<SupportedChainId>
+
+export function useMultiNetworkAddressBalances({ ownerAddress, tokenAddress }: useMultiNetworkAddressBalancesArgs) {
+  const [data, setData] = useState<AddressNetworkBalanceData>({})
+  const [error] = useState<string | null>(null)
+  const [loading, setLoading] = useState(false)
+
+  const { chainId: connectedChainId } = useWeb3React()
+  const feature_flag_multi_network_balances = useBaseFlag(FeatureFlag.multiNetworkBalances)
+
+  const handleBalance = useCallback(({ amount, chainId, tokenAddress }: handleBalanceArg) => {
+    if (!amount.greaterThan(0) || !tokenAddress) {
+      return
+    }
+    setData((data) => ({
+      ...data,
+      [chainId]: {
+        ...(data[chainId] ?? {}),
+        [tokenAddress]: amount,
+      },
+    }))
+  }, [])
+
+  useEffect(() => {
+    if (!ownerAddress || !tokenAddress) {
+      return
+    }
+    const isConnecteToTestnet = connectedChainId ? TESTNET_CHAIN_IDS.includes(connectedChainId) : false
+    setLoading(true)
+    const isNative = tokenAddress === 'NATIVE'
+    const promises: Promise<any>[] = []
+
+    const isWrappedNative = ALL_SUPPORTED_CHAIN_IDS.some(
+      (chainId) => WRAPPED_NATIVE_CURRENCY[chainId]?.address.toLowerCase() === tokenAddress.toLowerCase()
+    )
+
+    const chainsToCheck: SupportedChainId[] =
+      feature_flag_multi_network_balances === BaseVariant.Enabled
+        ? ALL_SUPPORTED_CHAIN_IDS
+        : isSupportedChain(connectedChainId)
+        ? [SupportedChainId.MAINNET, connectedChainId]
+        : [SupportedChainId.MAINNET]
+
+    chainsToCheck.forEach((chainId) => {
+      const isTestnet = testnetSet.has(chainId)
+      if ((isConnecteToTestnet && isTestnet) || !isTestnet) {
+        const provider = RPC_PROVIDERS[chainId]
+        if (isWrappedNative || isNative) {
+          const wrappedNative = WRAPPED_NATIVE_CURRENCY[chainId]
+          if (wrappedNative) {
+            promises.push(
+              new Promise(async (resolve) => {
+                try {
+                  const wrappedNativeContract = getContract(wrappedNative.address, WETH_ABI, provider) as Weth
+                  const balance = await wrappedNativeContract.balanceOf(ownerAddress, { blockTag: 'latest' })
+                  const amount = CurrencyAmount.fromRawAmount(wrappedNative, balance.toString())
+                  resolve(handleBalance({ amount, chainId, tokenAddress: wrappedNative.address.toLowerCase() }))
+                } catch (e) {}
```

Comment:
console.log(e) ?

Reply:
nah I think it's better to throw this one into the void, since I'm expecting there to be a lot of errors when querying addresses that don't exist on some networks